account_reconcile_rule/model/account_reconcile_rule.py

```diff
@@ -0,0 +1,168 @@
+            return False
+        currency = statement_line.currency_for_rules()
+        if currency not in self.currency_ids:
+            return False
+        amount_currency = statement_line.amount_currency
+        for move_line in move_lines:
+            if move_line.currency_id != statement_line.currency_id:
+                # use case not supported, no rule found
+                return False
+            amount_currency -= move_line.amount_currency
+
+        # amount in currency is the same, so the balance is
+        # a difference due to currency rates
+        if statement_line.currency_id.is_zero(amount_currency):
+            return self._balance_in_range(balance, currency)
+        return False
+
+    @api.multi
+    def is_valid(self, statement_line, move_lines, balance):
+        """ Returns True if a rule applies to a group of statement_line +
+        move lines.
+
+        This is the public method where the rule is evaluated whatever
+        its type is.  When a rule returns True, it means that it is a
+        candidate for the current reconciliation. The rule with the lowest
+        number in the ``sequence`` field is chosen.
+
+        :param statement_line: the line to reconcile
+        :param move_lines: the selected move lines for reconciliation
+        :param balance: the balance between the statement_line and the
+                        move_lines. It could be computed here but it is
+                        computed before to avoid to compute it for each
+                        rule when called on multiple rules.
+        """
+        self.ensure_one()
+        if self.rule_type == 'rounding':
+            return self._is_valid_balance(statement_line, balance)
+        elif self.rule_type == 'currency':
+            return self._is_valid_multicurrency(statement_line,
+                                                move_lines,
+                                                balance)
+
+    @api.model
+    def find_first_rule(self, statement_line, move_lines):
+        """ Find the rules that apply to a statement line and
+        a selection of move lines.
+
+        :param statement_line: the line to reconcile
+        :param move_lines: the selected move lines for reconciliation
+        """
+        balance = statement_line.amount
+        for move_line in move_lines:
+            balance += move_line.credit - move_line.debit
+
+        currency = statement_line.currency_for_rules()
+        if currency.is_zero(balance):
+            return self.browse()
+
+        rules = self.search([])
+        # return the first applicable rule
+        for rule in rules:
+            if rule.is_valid(statement_line, move_lines, balance):
+                return rule
+        return self.browse()
+
+    @api.model
+    @api.returns('account.reconcile.model')
+    def models_for_reconciliation(self, statement_line_id, move_line_ids):
+        """ Find the rule for the current reconciliation and returns the
```

Comment:
`"""Find reconcile models for given statement and move lines.`
explain "how" in the extra lines.